### PR TITLE
Add ANGLE shader translator fuzzer compile targets

### PR DESCRIPTION
**Pull request by kkinnunen-apple**
#### c90aee964745d4ce9443f3ed1e2e71f51eca323a
<pre>
Add ANGLE shader translator fuzzer compile targets
<a href="https://bugs.webkit.org/show_bug.cgi?id=278782">https://bugs.webkit.org/show_bug.cgi?id=278782</a>
<a href="https://rdar.apple.com/134846437">rdar://134846437</a>

Reviewed by David Kilzer.

Adds fuzz target that are not built by default WebKit build.
Build by
  make release SCHEME=&quot;Fuzzers (ANGLE)&quot; ASAN=YES LIBFUZZER=YES

ANGLEShaderTranslator
 - fuzz target
 - links statically to libtranslator (see below)

Adds the existing metal_translator_fuzzer to this &quot;Fuzzers (ANGLE)&quot;
aggregate target, too.

Adds fuzzing tools that are not built by default WebKit build.
Build by
  make release SCHEME=&quot;Tools (ANGLE)&quot; ASAN=NO LIBFUZZER=NO

ANGLEShaderTranslatorCoverage
  - XCTest project to run the corpus with the compile flags of
    ANGLEShaderTranslator
  - Can be used to verify the coverage of the corpus

ANGLEShaderTranslatorStats
  - Used to analyze RSS increases when corpus is run
  - Run the corpus with the compile flags of
    ANGLEShaderTranslator

Adds a static library that is not built by default WebKitBuild

libtranslator
  - Compiles translator again, to be linked statically to the fuzzer
  - In later commits, may be linked to libANGLE-shared. Not done
    in this commit to avoid breakage.

* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/xcshareddata/xcschemes/Fuzzers (ANGLE).xcscheme: Added.
* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/xcshareddata/xcschemes/Tools (ANGLE).xcscheme: Added.
* Source/ThirdParty/ANGLE/Configurations/ANGLETranslatorFuzzer.xcconfig: Added.
* Source/ThirdParty/ANGLE/Configurations/ANGLETranslatorFuzzerCoverage.xcconfig: Added.
* Source/ThirdParty/ANGLE/Configurations/ANGLETranslatorFuzzerDumpTestCase.xcconfig: Added.
* Source/ThirdParty/ANGLE/Configurations/ANGLETranslatorFuzzerStats.xcconfig: Added.
* Source/ThirdParty/ANGLE/Configurations/translator.xcconfig: Added.
* Source/ThirdParty/ANGLE/WebKit/TranslatorFuzzer.cpp: Added.
(sh:: const):
(sh::mutateOptions):
(sh::initializeValidFuzzerOptions):
(sh::mutate):
(sh::initializeTranslators):
(sh::getTranslator):
(sh::initializeFuzzer):
(filterOptions):
(LLVMFuzzerCustomMutator):
(LLVMFuzzerTestOneInput):
* Source/ThirdParty/ANGLE/WebKit/TranslatorFuzzerCoverage.mm: Added.
(LLVMFuzzerMutate):
(-[TranslatorFuzzerCoverage setUp]):
(-[TranslatorFuzzerCoverage tearDown]):
(-[TranslatorFuzzerCoverage testFuzzerCorpusCoverage]):
* Source/ThirdParty/ANGLE/WebKit/TranslatorFuzzerCoverageInfo.plist: Added.
* Source/ThirdParty/ANGLE/WebKit/TranslatorFuzzerDumpTestCase.cpp: Added.
(LLVMFuzzerMutate):
* Source/ThirdParty/ANGLE/WebKit/TranslatorFuzzerStats.cpp: Added.
(LLVMFuzzerMutate):
(getRSSKB):
(main):
* Source/ThirdParty/ANGLE/WebKit/TranslatorFuzzerSupport.h: Added.
(GLSLDumpHeader::GLSLDumpHeader):
(GLSLDumpHeader::write):
* Source/ThirdParty/ANGLE/WebKit/webgl-glsl.dict: Added.

Canonical link: <a href="https://commits.webkit.org/283977@main">https://commits.webkit.org/283977@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/eb4c862f525df6e720c5d6289cf5e36df05556ab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65071 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44438 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17685 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69095 "Built successfully") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67189 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52221 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15959 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52271 "Passed tests") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68137 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41089 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56321 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32893 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37758 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13692 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14553 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59670 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14030 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70800 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9023 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13511 "Found 1 new test failure: imported/w3c/web-platform-tests/media-source/mediasource-activesourcebuffers.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59599 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9055 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56381 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59848 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15100 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7448 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1127 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40250 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41327 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42508 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41071 "Built successfully") | | |

### CODE REVIEW

File: Configurations/Sanitizers.xcconfig [hunk 1]
```diff
@@ -50,7 +54,7 @@ WK_SANITIZER_WARNING_CFLAGS = $(WK_STATIC_ANALYZER_WARNING_CFLAGS_$(RUN_CLANG_ST
 // FIXME: Tell Xcode not to compile host-side tools with sanitizers enabled (see <rdar://99386433>).
 // Workaround is to use `WK_SANITIZER_DISALLOWED = YES`.
 
-WK_ANY_SANITIZER_ENABLED = $(WK_AND_$(WK_OR_$(ENABLE_ADDRESS_SANITIZER)_$(WK_OR_$(ENABLE_THREAD_SANITIZER)_$(WK_OR_$(ENABLE_UNDEFINED_BEHAVIOR_SANITIZER)_$(ENABLE_LIBFUZZER))))_$(WK_NOT_$(WK_SANITIZER_DISALLOWED)));
+WK_ANY_SANITIZER_ENABLED = $(WK_AND_$(WK_OR_$(ENABLE_ADDRESS_SANITIZER)_$(WK_OR_$(ENABLE_THREAD_SANITIZER)_$(WK_OR_$(ENABLE_UNDEFINED_BEHAVIOR_SANITIZER)_$(ENABLE_LIBFUZZER)))));
```
Comment (ddkilzer) [inline Configurations/Sanitizers.xcconfig]: I don't understand why using `$(WK_NOT_$(WK_SANITIZER_DISALLOWED))` as part of `WK_ANY_SANITIZER_ENABLED` didn't work as-is?

This seems like the wrong change since you didn't remove the `$(WK_AND_$(...))` when you removed the `_$(WK_NOT_$(WK_SANITIZER_DISALLOWED))`.

File: Source/ThirdParty/ANGLE/WebKit/TranslatorFuzzer.cpp [hunk 2]
```diff
@@ -0,0 +1,271 @@
+// Copyright 2016 The ANGLE Project Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (ddkilzer) [inline Source/ThirdParty/ANGLE/WebKit/TranslatorFuzzer.cpp]: This should also have an Apple copyright.  I think there are Apple changes here that aren't part of the baseline fuzzer.

File: Source/ThirdParty/ANGLE/Configurations/ANGLETranslatorFuzzerDynamic.xcconfig [hunk 3]
```diff
@@ -23,4 +23,3 @@
 
 #include "BaseTarget.xcconfig"
 
-OTHER_CFLAGS = $(inherited) -DANGLE_TRANSLATOR_FUZZER_METAL_ONLY;
```
Comment (ddkilzer) [inline Source/ThirdParty/ANGLE/Configurations/ANGLETranslatorFuzzerDynamic.xcconfig]: Where did this go?  Or why was it removed?  Is it no longer needed?

File: Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -3985,6 +5246,753 @@
 			};
 			name = Production;
 		};
+		7BFBD7AF2C7DE55300BFCA94 /* Debug */ = {
+			isa = XCBuildConfiguration;
+			baseConfigurationReference = 7BFBD7E02C7DE7C700BFCA94 /* ANGLETranslatorFuzzerStaticAssert.xcconfig */;
+			buildSettings = {
+				ALWAYS_SEARCH_USER_PATHS = NO;
```
Comment (emw-apple) [inline Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj]: Could you please remove the default build settings in all these new XCBuildConfiguration objects? They should be unnecessary to build, as long as each target has a `Base.xcconfig` in its includes and defines `PRODUCT_NAME` in xcconfigs.

File: Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -3879,7 +4592,545 @@
...
+		};
+		7BFBD7A72C7DE55300BFCA94 /* Sources */ = {
+			isa = PBXSourcesBuildPhase;
+			buildActionMask = 2147483647;
+			files = (
+				7BFBD8DC2C7E025300BFCA94 /* AddExplicitTypeCasts.cpp in Sources */,
+				7BFBD7EC2C7DE96D00BFCA94 /* aligned_memory.cpp in Sources */,
+				7BFBD7E92C7DE95100BFCA94 /* angle_version_info.cpp in Sources */,
+				7BFBD8682C7E01D100BFCA94 /* angleutils.cpp in Sources */,
+				7BFBD8A92C7E023200BFCA94 /* AstHelpers.cpp in Sources */,
```
Comment (ddkilzer) [inline Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj]: Doesn't this basically duplicate (three times!) all of the source files used to compile libANGLE-shared.dylib?

It's a shame we have to duplicate targets instead of re-using them.

That's why I created `libJavaScriptCore.a` in the `JavaScriptCore` project—so that I didn't have to create a separate target to build all the source files twice.

The same thing could be done here, although it's a bit gnarly to get it correct, so this is probably the best option for now, but it's going to break frequently (unless you commit to testing the build every time ANGLE is updated from upstream).

TL;DR: No need to make a change for this PR, but long term I'd like to see a `libANGLE.a` output (intermediate build product) that is then linked into `libANGLE-shared.dylib` if you need both static and dynamic linking of the code.

File: Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -3985,6 +5246,753 @@
 			};
 			name = Production;
 		};
+		7BFBD7AF2C7DE55300BFCA94 /* Debug */ = {
+			isa = XCBuildConfiguration;
+			baseConfigurationReference = 7BFBD7E02C7DE7C700BFCA94 /* ANGLETranslatorFuzzerStaticAssert.xcconfig */;
+			buildSettings = {
+				ALWAYS_SEARCH_USER_PATHS = NO;
+				ASSETCATALOG_COMPILER_GENERATE_SWIFT_ASSET_SYMBOL_EXTENSIONS = YES;
+				CLANG_ANALYZER_NONNULL = YES;
```
Comment (ddkilzer) [inline Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj]: You need to manually delete all of these `buildSettings` since there is a `*.xcconfig` file to use (except maybe `PRODUCT_NAME`, unless that's in the xcconfig file, too).

Probably easiest just to open the `project.pbxproj` in a text editor and delete the lines (as long as the text editor doesn't try to change tabs to spaces or make other whitespace changes without explicit action).

This is the worst part about adding new targets if you don't copy them—Xcode re-adds all these settings after folks painstakingly moved them out to xcconfig files earlier.

File: Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -3985,6 +5246,753 @@
...
+			};
+			name = Debug;
+		};
+		7BFBD7B02C7DE55300BFCA94 /* Release */ = {
+			isa = XCBuildConfiguration;
+			baseConfigurationReference = 7BFBD7E02C7DE7C700BFCA94 /* ANGLETranslatorFuzzerStaticAssert.xcconfig */;
+			buildSettings = {
+				ALWAYS_SEARCH_USER_PATHS = NO;
+				ASSETCATALOG_COMPILER_GENERATE_SWIFT_ASSET_SYMBOL_EXTENSIONS = YES;
+				CLANG_ANALYZER_NONNULL = YES;
```
Comment (ddkilzer) [inline Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:5318]: Ditto.

File: Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -3985,6 +5246,753 @@
...
+			};
+			name = Release;
+		};
+		7BFBD7B12C7DE55300BFCA94 /* Production */ = {
+			isa = XCBuildConfiguration;
+			baseConfigurationReference = 7BFBD7E02C7DE7C700BFCA94 /* ANGLETranslatorFuzzerStaticAssert.xcconfig */;
+			buildSettings = {
+				ALWAYS_SEARCH_USER_PATHS = NO;
+				ASSETCATALOG_COMPILER_GENERATE_SWIFT_ASSET_SYMBOL_EXTENSIONS = YES;
+				CLANG_ANALYZER_NONNULL = YES;
```
Comment (ddkilzer) [inline Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:5375]: Ditto.

File: Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -3985,6 +5246,753 @@
...
+			};
+			name = Release;
+		};
+		7BFBD7DD2C7DE76100BFCA94 /* Production */ = {
+			isa = XCBuildConfiguration;
+			baseConfigurationReference = 7BFBD7E22C7DE7C700BFCA94 /* ANGLETranslatorFuzzerTool.xcconfig */;
+			buildSettings = {
+				ALWAYS_SEARCH_USER_PATHS = NO;
+				ASSETCATALOG_COMPILER_GENERATE_SWIFT_ASSET_SYMBOL_EXTENSIONS = YES;
+				CLANG_ANALYZER_NONNULL = YES;
```
Comment (ddkilzer) [inline Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj]: Ditto for the other 11 target+configuration combos above.  :)

**Review by ddkilzer — CHANGES_REQUESTED** [review]
r- to make suggested changes, and make changes we discussed in the Slack channel.

File: Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj (hunk 4 above)
Comment (ddkilzer) [reply Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj]: Oops, I also mentioned this in my review (which I started before lunch, then just finished).  It's the same as all of the "Ditto" comments, plus the [original comment](https://github.com/WebKit/WebKit/pull/32819#discussion_r1737252604).

**Review by ddkilzer — APPROVED** [review]
r=me
